### ISSUE TITLE
remote cache hits but is not used!

### ISSUE THREAD

**Issue by emranbm** [bazelbuild/bazel#10011]
### Description of the problem / feature request:

Bazel remote cache hits, but the build runs completely before the cache hit!

### Bugs: what's the simplest, easiest way to reproduce this bug? Please provide a minimal example if possible.

Reproducer docker image:
```dockerfile
FROM ubuntu:16.04

WORKDIR /opt/

COPY bazel-0.29.1-installer-linux-x86_64.sh .
COPY src/ src/
COPY distdir/ /opt/distdir/

RUN apt-get update
RUN apt-get install unzip
RUN apt-get install -y g++
RUN ./bazel-0.29.1-installer-linux-x86_64.sh --user \
&& rm ./bazel-0.29.1-installer-linux-x86_64.sh \
&& ln -s /root/bin/bazel /usr/bin/bazel \
&& bazel version
RUN touch WORKSPACE
```
`src` directory:
```
src/
├── BUILD    >>> defines `main` target
└── main.cc  >>> simple hello world
```
docker-compose.yml:
```yml
version: "3.7"
services:
  cache-test:
    build: .
    command: "bazel build -s src:main --distdir=/opt/distdir/ --remote_cache=http://cacheserver:9090"
```
Now just run `docker-compose up` and the log seems like:
```
Attaching to cache_cache-test_1
cache-test_1  | Starting local Bazel server and connecting to it...
cache-test_1  | INFO: Invocation ID: 4ea41648-9c33-411e-8567-f45417954581
cache-test_1  | Loading: 
cache-test_1  | Loading: 0 packages loaded
cache-test_1  | Analyzing: target //src:main (1 packages loaded, 0 targets configured)
cache-test_1  | INFO: Analyzed target //src:main (13 packages loaded, 68 targets configured).
cache-test_1  | INFO: Found 1 target...
cache-test_1  | [0 / 6] [Prepa] BazelWorkspaceStatusAction stable-status.txt ... (3 actions, 0 running)
cache-test_1  | SUBCOMMAND: # //src:main [action 'Compiling src/main.cc']
cache-test_1  | (cd /root/.cache/bazel/_bazel_root/68ab1268abb7e5f5688810968df97d6b/execroot/__main__ && \
cache-test_1  |   exec env - \
cache-test_1  |     PATH=/usr/local/sbin:/usr/local/bin:/usr/sbin:/usr/bin:/sbin:/bin \
cache-test_1  |     PWD=/proc/self/cwd \
cache-test_1  |   /usr/bin/gcc -U_FORTIFY_SOURCE -fstack-protector -Wall -Wunused-but-set-parameter -Wno-free-nonheap-object -fno-omit-frame-pointer '-std=c++0x' -MD -MF bazel-out/k8-fastbuild/bin/src/_objs/main/main.pic.d '-frandom-seed=bazel-out/k8-fastbuild/bin/src/_objs/main/main.pic.o' -fPIC -iquote . -iquote bazel-out/k8-fastbuild/bin -iquote external/bazel_tools -iquote bazel-out/k8-fastbuild/bin/external/bazel_tools -fno-canonical-system-headers -Wno-builtin-macro-redefined '-D__DATE__="redacted"' '-D__TIMESTAMP__="redacted"' '-D__TIME__="redacted"' -c src/main.cc -o bazel-out/k8-fastbuild/bin/src/_objs/main/main.pic.o)
cache-test_1  | SUBCOMMAND: # //src:main [action 'Linking src/main']
cache-test_1  | (cd /root/.cache/bazel/_bazel_root/68ab1268abb7e5f5688810968df97d6b/execroot/__main__ && \
cache-test_1  |   exec env - \
cache-test_1  |     PATH=/usr/local/sbin:/usr/local/bin:/usr/sbin:/usr/bin:/sbin:/bin \
cache-test_1  |     PWD=/proc/self/cwd \
cache-test_1  |   /usr/bin/gcc @bazel-out/k8-fastbuild/bin/src/main-2.params)
cache-test_1  | Target //src:main up-to-date:
cache-test_1  |   bazel-bin/src/main
cache-test_1  | INFO: Elapsed time: 2.855s, Critical Path: 0.19s
cache-test_1  | INFO: 2 processes: 2 remote cache hit.
cache-test_1  | INFO: Build completed successfully, 6 total actions
cache-test_1  | INFO: Build completed successfully, 6 total actions
```
As you see, the whole compilation and linkage processes ran and at the end, the log says two processes hit the cache!
But as you see, the whole processes are ran and the `Elapsed time` is almost the same as `Elapsed time` without cache (i.e. with --remote_cache flag omitted)

### What operating system are you running Bazel on?

ubuntu 16.04

### What's the output of `bazel info release`?

release 0.29.1

###  Have you found anything relevant by searching the web?
[Here](https://docs.bazel.build/versions/master/remote-caching-debug.html#troubleshooting-cache-hits) is about how to debug if not getting _expected hit rate_. However, my problem is not about the hit rate, but using the hit results.

**Comment by emranbm** [comment]
Same issue with bazel 1.0.0
Any help?

**Comment by buchgr** [comment]
Is the problem that you are seeing `-s` displaying command lines? This is expected and in no way indicates the remote cache not being used.

Please re-open if that's not the issue.

**Comment by emranbm** [comment]
Thanks for your reply. But, as I mentioned, It's not just about seeing the command.:
> But as you see, the whole processes are ran and the Elapsed time is almost the same as Elapsed time without cache (i.e. with --remote_cache flag omitted)

**Comment by buchgr** [comment]
```
INFO: 2 processes: 2 remote cache hit.
```

This line means that both processes / actions were a remote cache hit. Is this what you were unsure about?

**Comment by emranbm** [comment]
Oh, we found the mistake!
We had written a `genrule` like below, and were wondering why the `echo ` command always runs.
But then, to be sure that the command is really being executed, we added the second echo command (which prints the timestamp), and surprisingly discovered that the remote cache not only caches the output files, but also the stdout/stderr of the executed rule.
```
genrule(
name = "main",
srcs = ["input.txt"],
outs = ["output.txt"],
cmd = """
echo From inside genrule
echo This cmd executed at: `date`
cp src/input.txt $(RULEDIR)/output.txt
"""
)
```
So the problem was not about the remote cache, but what it prints!
To prevent such a misunderstanding, seems a good idea to mention this behavior (printing cached stdout...) in remote cache docs.
Another suggestion would be printing the cached stdout/stderr in a different distinguishable format during the build process.
